modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultSemanticAnalysisPhase.java

```diff
@@ -2055,58 +2061,165 @@ public void visitRegex(ERegex userRegexNode, SemanticScope semanticScope) {
-                    new IllegalArgumentException("invalid regular expression: " +
-                            "could not compile regex constant [" + pattern + "] with flags [" + flags + "]", pse));
+            
+            Map<String, String> patternBank = semanticScope.getScriptScope().getCompilerSettings().getGrokPatternBank();
+            MatcherWatchdog watchdog = semanticScope.getScriptScope().getGrokWatchdog();
+
+            /*
+             * Shift from the "no permissions" compilation context to one that
+             * has permission to read stuff from the classpath so that jcodings,
+             * a dependency of grok, can read what it needs from the classpath. 
+             */
+            return AccessController.doPrivileged(new PrivilegedAction<Grok>() {
+                @Override
+                public Grok run() {
+                    // Try to collect warnings up front and refuse to compile the expression if there are any
+                    List<String> warnings = new ArrayList<>();
+                    new Grok(patternBank, regex.getPattern(), watchdog, warnings::add).match("___nomatch___");
+                    if (false == warnings.isEmpty()) {
+                        throw regex.createError(
+                            new IllegalArgumentException(
+                                "Compiled grok pattern constant [" + regex.getPattern() + "] but emitted warnings: " + warnings
+                            )
+                        );
+                    }
+
+                    return new Grok(
+                        patternBank,
+                        regex.getPattern(),
+                        watchdog,
+                        w -> {
+                            throw new IllegalArgumentException("grok [" + regex.getPattern() + "] emitted a warning: " + w);
+                        }
+                    );
+                }
+            });
+        } catch (IllegalArgumentException e) {
+            Location errorLocation = new Location(
+                regex.getLocation().getSourceName(),
+                regex.getLocation().getOffset() + regex.getPatternStart()
+            );
+            throw errorLocation.createError(
+                new IllegalArgumentException("Could not compile grok pattern constant [" + regex.getPattern() + "]: " + e.getMessage(), e)
+            );
         }
+    }
 
-        semanticScope.putDecoration(userRegexNode, new ValueType(Pattern.class));
-        semanticScope.putDecoration(userRegexNode, new StandardConstant(constant));
+    private DissectParser compileDissect(ERegex regex) {
+        if (false == "".equals(regex.getFlags())) {
+            throw regex.createError(
+                new IllegalArgumentException("Dissect doesn't support flags but got [" + regex.getFlags() + "]")
+            );
+        }
+        try {
+            return new DissectParser(regex.getPattern(), null);
```

Comment:
Dissect is not a backtracking regex engine so we don't have to wrap the string or really do anything to prevent it running forever. Yay.

That `null` here is the "appendSeparator", a thing that'd be useful to configure but I have no where to configure it at the moment. Unlike grok's pattern bank, it is kind of something you'd want to configure per dissect constant. I dunno so I'm going to leave it as `null` which means "default" for now.